app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt

```diff
@@ -0,0 +1,126 @@
+package edu.stanford.bdh.engagehf.bluetooth.data.repository
+
+import androidx.health.connect.client.records.BloodPressureRecord
+import androidx.health.connect.client.records.HeartRateRecord
+import androidx.health.connect.client.records.WeightRecord
+import ca.uhn.fhir.context.FhirContext
+import ca.uhn.fhir.context.FhirVersionEnum
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Query
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import edu.stanford.healthconnectonfhir.Loinc
+import edu.stanford.healthconnectonfhir.ObservationToRecordMapper
+import edu.stanford.spezi.core.coroutines.di.Dispatching
+import edu.stanford.spezi.core.logging.speziLogger
+import edu.stanford.spezi.module.onboarding.invitation.await
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import org.hl7.fhir.r4.model.Observation
+import javax.inject.Inject
+
+internal class ObservationRepository @Inject constructor(
+    private val firestore: FirebaseFirestore,
+    private val firebaseAuth: FirebaseAuth,
+    private val observationToRecordMapper: ObservationToRecordMapper,
+    @Dispatching.IO private val ioDispatcher: CoroutineDispatcher,
+) {
+    private val logger by speziLogger()
+
+    private val heartRateCollection =
+        ObservationCollection("heartRateObservations", Loinc.HeartRate.CODE)
+
+    private val bodyWeightObservation =
+        ObservationCollection("bodyWeightObservations", Loinc.Weight.CODE)
+
+    private val bloodPressureCollection =
+        ObservationCollection("bloodPressureObservations", Loinc.BloodPressure.CODE)
+
+    private val jsonParser by lazy {
+        FhirContext.forCached(FhirVersionEnum.R4).newJsonParser()
+    }
+
+    private val gson by lazy {
+        Gson()
+    }
+
+    suspend fun saveObservations(observations: List<Observation>) {
+        withContext(ioDispatcher) {
+            runCatching {
+                val uid = firebaseAuth.currentUser?.uid
+                    ?: throw IllegalStateException("User not authenticated")
+                val batch = firestore.batch()
+
+                observations.forEach { observation ->
+                    val collectionName = getCollectionNameForObservation(observation)
+                    val json = jsonParser.encodeResourceToString(observation)
+                    val mapType = object : TypeToken<Map<String, Any>>() {}.type
+                    val data = gson.fromJson<Map<String, Any>>(json, mapType)
+
+                    val docRef = firestore.collection("users/$uid/$collectionName").document()
+                    batch.set(docRef, data)
+                }
+
+                batch.commit().await()
+            }.onFailure {
+                logger.e(it) { "Error while saving observations" }
+            }
+        }
+    }
+
+    private fun getCollectionNameForObservation(observation: Observation): String {
```

Comment:
If we pull collations, we should look into snapshot listeners to ensure that we always get the latest set of values in case something changes: https://firebase.google.com/docs/firestore/query-data/listen#kotlin+ktx

Reply:
Thanks for the tip At this point it seems like it might just be overhead because the new measurements only come from the app right now. However, I will be happy to try this out with the messages feature, as the new data comes from outside the app and it will make it much easier to stay up to date at all times and when we notice something.